presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/optimization/JdbcComputePushdown.java

```diff
@@ -132,32 +137,74 @@ public PlanNode visitFilter(FilterNode node, Void context)
                     jdbcFilterToSqlTranslator,
                     oldTableScanNode.getAssignments());
 
-            // TODO if jdbcExpression is not present, walk through translated subtree to find out which parts can be pushed down
-            if (!oldTableHandle.getLayout().isPresent() || !jdbcExpression.getTranslated().isPresent()) {
+            if (!oldTableHandle.getLayout().isPresent()) {
                 return node;
             }
 
+            Optional<JdbcExpression> translated = jdbcExpression.getTranslated();
+            // All filter can pushdown
+            if (translated.isPresent()) {
+                return createNewTableScanNode(oldTableScanNode, oldTableHandle, oldConnectorTable, translated);
+            }
+            else {
+                // Find out which parts can be pushed down
+                List<RowExpression> pushable = new ArrayList<>();
+                List<RowExpression> nonPushable = new ArrayList<>();
+                List<JdbcExpression> translatedExpressions = new ArrayList<>();
+
+                List<RowExpression> rowExpressions = LogicalRowExpressions.extractConjuncts(predicate);
+                for (RowExpression expression : rowExpressions) {
+                    TranslatedExpression<JdbcExpression> translatedExpression = translateWith(
+                            expression,
+                            jdbcFilterToSqlTranslator,
+                            oldTableScanNode.getAssignments());
+
+                    if (!translatedExpression.getTranslated().isPresent()) {
+                        nonPushable.add(expression);
+                    }
+                    else {
+                        translatedExpressions.add(translatedExpression.getTranslated().get());
+                        pushable.add(expression);
+                    }
+                }
+
+                // All filter can't pushdown
+                if (!nonPushable.isEmpty() && pushable.isEmpty()) {
+                    return node;
+                }
+                else { // Part of the filter can be pushed down
```

Comment:
- same; `else` is not needed; it's redundant
- no need to have the comment; it's self-explanatory